### ISSUE TITLE
server crash when a map is missing a dependence

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#2046]
When a map requires a dependence which is absent from the server, it crashes.
Expected behavior would be to skip for next map, or to refuse the map change beforehand.

**Comment by DolceTriade** [comment]
Our current map logic is very naive:
 - Scan all pakpaths for dpks prefixed with `map-`
 - Report them

This is because we do not actually load the map. I think the biggest thing is that we have no such concept as "commands failing". You run a command it either preforms an action or does nothing.

I think this might be non-trivial to fix, but certainly we should try to come up with a way of validating maps prior to diving headfirst into them :)

**Comment by illwieckz** [comment]
> I think this might be non-trivial to fix, but certainly we should try to come up with a way of validating maps prior to diving headfirst into them :)

But this would defeat what was just implemented to not have to actually read the dpk to produce map list.

This is probably non trivial, but it would be better to load the next map in rotation if the map load fail.

Map load failure can also come from other things than the absence of `.bsp` file, it can happens, for example, if the `.bsp` file is broken. So the best file validation is to actually load the file with the code expected to load it.

**Comment by slipher** [comment]
You can configure your server to start a new game if the sgame stops:
```
set common.watchdogTime 1  // to do it faster
set common.watchdogCmd "map plat23"
```

**Comment by slipher** [comment]
If we add my last message into the sample config, can we consider this fixed?

If not, what else would you want to happen?

**Comment by illwieckz** [comment]
Would the map rotation continue to the next step?

Like if map rotation is `antares, forlorn, broken, station15, spacetracks`, after antares and forlorn are loaded, once broken is attempted to be loaded, the watchdog loads plat23 and after plat23 station15 is loaded?

**Comment by slipher** [comment]
The map rotation state is stored as cvars, so the  rotation state will still be there. But we should double-check that it doesn't do `broken` again.

**Comment by DolceTriade** [comment]
Unless we make this the default or we can do something else that's sensible without having server admins manually set cvars, I don't think we can close this.

**Comment by slipher** [comment]
Well the default behavior for the server is to shutdown 60 seconds after an sgame drop. This makes sense as server hosters usually restart `daemonded` in a loop (since the engine can crash too). Maybe that's better than trying to start a map anyway, in case an sgame crash is linked to some kind of persistent state in the engine.

For those who prefer the engine restart we also have `common.shutdownOnDrop` which can make it exit immediately.

**Comment by DolceTriade** [comment]
I think if we can find a way to skip non-loadable maps, that would be ideal. As illwieckz mentions, we don't want to get into a situation where we continuously crash at the same invalid map if we restart.

While most admins do start daemonded in a loop, it doesn't seem like we should rely on that behavior as part of our solution to deal with invalid map dpks.

**Comment by illwieckz** [comment]
Also engine server restart disconnects the client, which isn't a good experience, especially since we don't have mechanism for the game to automatically try to reconnect on disconnection, waiting for the server to be up again.

**Comment by slipher** [comment]
All clients are disconnected on any sgame drop. ServerApplication::OnDrop calls SV_Shutdown which tells all the clients to disconnect.

**Comment by DolceTriade** [comment]
Yes, my understanding is that we want to avoid a drop for clients if possible

**Comment by bmorel** [comment]
expected behavior would clearly be that server tries it's best effort to not loose clients, and if it have, then to send them a decent error message.
If the server boots anew or goes on next maprotation map until a valid one is found, it's fine. Still needs to send to clients the info of the list of maps (in a single window, preferably) which failed, so that a bug report can be done, instead of a silent disconnect.
IMO we should avoid those silent bugs as much as possible: they do not help us to fix bugs, far from it. Maybe we could somehow have a way to have daemon and daemoneded to talk together for errors, stuff stored in a reserved memory area, that is only readable by the nexes?
I don't know.
